// @recallai/desktop-sdk/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.init = init;
exports.shutdown = shutdown;
exports.dumpAXTree = dumpAXTree;
exports.startRecording = startRecording;
exports.stopRecording = stopRecording;
exports.pauseRecording = pauseRecording;
exports.resumeRecording = resumeRecording;
exports.uploadRecording = uploadRecording;
exports.prepareDesktopAudioRecording = prepareDesktopAudioRecording;
exports.requestPermission = requestPermission;
exports.testUnexpectedShutdown = testUnexpectedShutdown;
exports.addEventListener = addEventListener;
const path = __importStar(require("path"));
const child_process_1 = require("child_process");
const readline = __importStar(require("node:readline"));
const node_stream_1 = require("node:stream");
const fs = __importStar(require("node:fs"));
const os = __importStar(require("node:os"));
const uuid_1 = require("uuid");
// HACK: This is needed because of https://github.com/electron/electron/issues/6262
// Search for the unpacked directory first, since fs.existsSync actually reports
// the "virtual" files in the .asar as being present
let exe_paths = [];
if (process.platform === "darwin") {
    exe_paths.push(path.join(__dirname, "desktop_sdk_macos_exe").replace('app.asar', 'app.asar.unpacked'));
    exe_paths.push(path.join(__dirname, "desktop_sdk_macos_exe"));
}
else if (process.platform === "win32") {
    exe_paths.push(path.join(__dirname, "agent-windows.exe").replace('app.asar', 'app.asar.unpacked'));
    exe_paths.push(path.join(__dirname, "agent-windows.exe"));
}
let proc;
let rlData;
let rlStdout;
let rlStderr;
const listeners = [];
const lastErrorOut = [];
const pendingCommands = {};
let lastOptions;
let remainingAutomaticRestarts = 10;
let unexpectedShutdown = false;
let logBuffer = [];
let logIndex = 0;
function appendError(error) {
    lastErrorOut.push(error);
    if (lastErrorOut.length > 100) {
        lastErrorOut.shift();
    }
}
function flushLogBuffer() {
    const buf = logBuffer.slice(); // just here for the copy
    logBuffer = [];
    logIndex = 0;
    // we use a for... in... here because it does not loop over the sparse (undefined) entries
    // if you doubt this, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in
    // and know that Nick and Tucker spent at least 10 minutes debating this
    for (const idx in buf) {
        const { level, log, echo } = buf[idx];
        doLog(level, [log], echo);
    }
}
function isPlainObject(value) {
    return (value !== null &&
        typeof value === "object" &&
        Object.getPrototypeOf(value) === Object.prototype);
}
/**
 * JSON.stringify with alphabetically sorted object keys (deep).
 * - Preserves array order
 * - Leaves non-plain objects (Date, Map, Set, class instances with toJSON, etc.) intact
 * - Detects circular references like JSON.stringify and throws
 */
function sortedStringify(value, replacer, space) {
    const seen = new WeakSet();
    function normalize(v) {
        if (v === null || typeof v !== "object")
            return v;
        const obj = v;
        if (seen.has(obj)) {
            throw new TypeError("Converting circular structure to JSON");
        }
        seen.add(obj);
        if (Array.isArray(v)) {
            const out = v.map(normalize);
            seen.delete(obj);
            return out;
        }
        if (isPlainObject(v)) {
            const src = v;
            const out = {};
            for (const key of Object.keys(src).sort()) {
                out[key] = normalize(src[key]);
            }
            seen.delete(obj);
            return out;
        }
        // Non-plain object: keep as-is so built-in toJSON/serialization applies.
        seen.delete(obj);
        return v;
    }
    return JSON.stringify(normalize(value), replacer, space);
}
async function doLog(level, log, echo = true) {
    try {
        const levelMap = {
            "info": "log",
            "warning": "warn",
            "error": "error"
        };
        if (echo) {
            const consoleMethod = console[levelMap[level] ?? "log"];
            consoleMethod(...log);
        }
        const logPayload = {
            log: log.join(" "),
            level: level,
            echo
        };
        const idx = logIndex++;
        logBuffer[idx] = logPayload;
        if (proc && proc.exitCode === null) {
            await sendCommand("log", logPayload);
            delete logBuffer[idx];
        }
    }
    catch (e) {
        console.error("Failed to send log to Desktop SDK", e.stack, e.message);
    }
}
function log(...log) {
    doLog("info", log);
}
function logWarning(...log) {
    doLog("warning", log);
}
function logError(...log) {
    doLog("error", log);
}
function emitEvent(type, payload) {
    if (type !== "upload-progress" && type !== "realtime-event" && type !== "log") {
        doLog("info", ["Receiving event: " + type + " | " + sortedStringify(payload)], false);
    }
    for (const listener of listeners) {
        if (listener.type === type) {
            listener.callback(payload);
        }
    }
}
function flushPendingCommands(err) {
    Object.keys(pendingCommands).forEach(commandId => {
        pendingCommands[commandId].reject(err);
        delete pendingCommands[commandId];
    });
}
function startProcess() {
    if (proc && proc.exitCode === null) {
        logError("Desktop SDK: Trying to start process while it is already started");
        return;
    }
    const exe_path = exe_paths.find(fs.existsSync);
    if (!exe_path) {
        logError(`Desktop SDK: Couldn't launch! This is likely an issue with the build tool you're using.`);
        for (const exe_path of exe_paths) {
            logError("Tried:", exe_path);
        }
        return;
    }
    let envExtra = {};
    if (process.platform === "win32" && process.env.GLOBAL_GST_RECALL !== "1") {
        envExtra["GST_PLUGIN_PATH"] = path.join(path.dirname(exe_path), "gstreamer-1.0");
    }
    if (process.platform === "darwin") {
        try {
            const electron = require('electron');
            envExtra["EXE_NAME"] = electron?.app?.getBundleID?.();
        }
        catch { }
    }
    if (!envExtra["EXE_NAME"]) {
        envExtra["EXE_NAME"] = path.basename(process.execPath);
    }
    const gst_dump_dir = process.env.RECALLAI_DESKTOP_SDK_DEV ? path.join(os.tmpdir(), "gst.nocommit") : os.tmpdir();
    if (!fs.existsSync(gst_dump_dir)) {
        try {
            fs.mkdirSync(gst_dump_dir);
        }
        catch (e) {
            logError("Failed to create gst dump dir:", String(e));
        }
    }
    proc = (0, child_process_1.spawn)(exe_path, {
        stdio: (process.platform === "darwin" ? ["pipe", "pipe", "pipe", "pipe"] : "pipe"),
        env: {
            GST_DEBUG: "2,transcriber:4,rtewebsocketsink:4,rtewebhooksink:4,audiomixer:4,galleryview:4,dsdks3sink:5",
            GST_DEBUG_DUMP_DOT_DIR: gst_dump_dir,
            RUST_BACKTRACE: "1",
            // "DYLD_INSERT_LIBRARIES":"/opt/homebrew/lib/libjemalloc.dylib",
            // "MALLOC_CONF":"prof:true,prof_active:true,prof_prefix:jeprof",
            // "DYLD_FORCE_FLAT_NAMESPACE":"1",
            ...envExtra,
        }
    });
    if (process.platform === "darwin") {
        const fd3 = proc.stdio[3];
        if (fd3 && fd3 instanceof node_stream_1.Readable) {
            rlData = readline.createInterface({ input: fd3, crlfDelay: Infinity });
        }
    }
    rlStdout = readline.createInterface({ input: proc.stdout, crlfDelay: Infinity });
    rlStderr = readline.createInterface({ input: proc.stderr, crlfDelay: Infinity });
    let onLine = (line) => {
        if (line?.startsWith("recall_ai_command|")) {
            try {
                const data = JSON.parse(line.substring(18));
                switch (data.type) {
                    case "event":
                        if (data.event) {
                            const event = JSON.parse(data.event);
                            emitEvent(event.type, event.payload);
                        }
                        break;
                    case "response":
                        const pendingCommand = pendingCommands[data.commandId];
                        if (pendingCommand) {
                            if (data.status === "success") {
                                pendingCommand.resolve(data.result);
                            }
                            else {
                                pendingCommand.reject(new Error(data.result));
                            }
                            delete pendingCommands[data.commandId];
                        }
                        break;
                }
            }
            catch (err) {
                logError("Desktop SDK: Failed to parse incoming data:", String(err));
            }
        }
        else {
            if (process.env.RECALLAI_DESKTOP_SDK_DEV) {
                console.log(line);
            }
        }
    };
    rlData?.on('line', onLine);
    rlStdout.on('line', onLine);
    rlStderr.on('line', (line) => {
        appendError(line);
        if (process.env.RECALLAI_DESKTOP_SDK_DEV) {
            console.error(line);
        }
    });
    proc.on('error', (error) => {
        flushPendingCommands(new Error(`Process error: ${error.message}`));
        emitEvent('error', {
            type: 'process',
            message: `The Desktop SDK server process has failed to start or exited improperly.`
        });
        logError(`Desktop SDK: Process error: ${error.message}`);
    });
    proc.on('close', async (code, signal) => {
        flushPendingCommands(new Error(`Process exited with code ${code}, signal ${signal}.`));
        emitEvent('shutdown', { code: code ?? 0, signal: signal ?? '' });
        rlStdout?.close();
        rlStderr?.close();
        rlStdout = null;
        rlStderr = null;
        if (code === 0 || signal === 'SIGINT') {
            return;
        }
        logError(`Desktop SDK: Process exited with code ${code}, signal ${signal}`);
        emitEvent('error', {
            type: 'process',
            message: "The Desktop SDK server process exited unexpectedly."
        });
        proc = null;
        unexpectedShutdown = true;
        if (lastOptions.restartOnError && remainingAutomaticRestarts > 0) {
            remainingAutomaticRestarts--;
            logError(`Automatically restarting Desktop SDK due to unexpected exit! Automatic restarts left: ${remainingAutomaticRestarts}`);
            logError(`Last error output from Desktop SDK:\n${lastErrorOut.join("\n")}`);
            lastErrorOut.length = 0;
            doInit(lastOptions);
        }
    });
}
function sendCommand(command, params = {}) {
    return new Promise((resolve, reject) => {
        if (!proc || !proc.stdin) {
            reject(new Error("The Desktop SDK is not started; call `init` to start it."));
            return;
        }
        const commandId = (0, uuid_1.v4)();
        pendingCommands[commandId] = { resolve, reject };
        const payload = {
            command,
            commandId,
            params
        };
        const payloadStr = JSON.stringify(payload);
        proc.stdin.write(payloadStr + "\n");
        if (command !== "log") {
            doLog("info", ["Sending command: " + payloadStr], false);
        }
    });
}
async function doInit(options) {
    startProcess();
    if (unexpectedShutdown) {
        logError("Desktop SDK: Recovered from unexpected shutdown");
        unexpectedShutdown = false;
    }
    await sendCommand("init", { config: JSON.stringify(options) });
    flushLogBuffer();
}
async function init(options) {
    if (process.platform !== "darwin" && process.platform !== "win32") {
        throw new Error(`Platform ${process.platform} is not supported by Desktop SDK`);
    }
    const { api_url, apiUrl, dev } = options;
    options.api_url = api_url ?? apiUrl ?? "https://api.recall.ai";
    if (!dev && (!options.api_url || !options.api_url.startsWith("https"))) {
        throw new Error(`apiUrl must be an https url, got: ${options.api_url}`);
    }
    if (options.restartOnError === undefined) {
        options.restartOnError = true;
    }
    lastOptions = options;
    await doInit(options);
    return null;
}
async function shutdown() {
    const result = await sendCommand("shutdown");
    if (proc) {
        const currentProc = proc;
        setTimeout(() => {
            if (!currentProc.killed) {
                currentProc.kill();
            }
        }, 5000);
    }
    return result;
}
function dumpAXTree(procName) {
    return sendCommand("dumpAXTree", { procName });
}
function startRecording(config) {
    return sendCommand("startRecording", { config: JSON.stringify(config) });
}
function stopRecording({ windowId }) {
    return sendCommand("stopRecording", { windowId });
}
function pauseRecording({ windowId }) {
    return sendCommand("pauseRecording", { windowId });
}
function resumeRecording({ windowId }) {
    return sendCommand("resumeRecording", { windowId });
}
function uploadRecording({ windowId }) {
    return sendCommand("uploadRecording", { windowId });
}
function prepareDesktopAudioRecording() {
    return sendCommand("prepareDesktopAudioRecording");
}
function requestPermission(permission) {
    return sendCommand("requestPermission", { permission });
}
function testUnexpectedShutdown() {
    return sendCommand("unexpectedShutdown", {});
}
function addEventListener(type, callback) {
    listeners.push({ type, callback });
}
const RecallAiSdk = {
    init,
    shutdown,
    startRecording,
    stopRecording,
    pauseRecording,
    resumeRecording,
    uploadRecording,
    prepareDesktopAudioRecording,
    requestPermission,
    addEventListener,
    testUnexpectedShutdown
};
exports.default = RecallAiSdk;
